# 1.Divide Region



In [1]:
# import libraries
import pandas as pd
import numpy as np
import csv
import matplotlib.pyplot as plt

In [2]:
ls

Country_religion/           WHO_policy_timeline_edited/
HFA_426_EN.csv              dataprocessed/
Policies.ipynb              policy_timeline.ipynb
Policies_tristan.ipynb      projectdata/
README.md*                  report/


In [12]:
#file_path_definitions
Age_limits_CSV_FILE_PATH = 'projectdata/Age_limits.csv'
Public_use_CSV_FILE_PATH = 'projectdata/Public_use.csv'
Liscence_retail_CSV_FILE_PATH = 'projectdata/Licence_retail.csv'  # bool influence the retail sales first
Liscence_production_CSV_FILE_PATH = 'projectdata/Licence_production.csv'
Tax_CSV_FILE_PATH = 'projectdata/Tax.csv'   # bool
Community_actions_CSV_FILE_PATH = 'projectdata/Community_action.csv'  #  bool
Health_warning_CSV_FILE_PATH = 'projectdata/Health_warning.csv'
Comsumption_CSV_FILE_PATH = 'projectdata/Consumption.csv'
Countries_list_PATH = 'dataprocessed/Country_list_final.csv'

In [13]:
%%javascript
IPython.OutputArea.auto_scroll_threshold = 9999;

<IPython.core.display.Javascript object>

1. We have divided all the countries in the database into eight parts according to the United Nations geographical division criteria.
2. This is because we take into account that countries in the same region have similar political conditions, living habits and cultural background, such as countries in the EU, while countries in different regions are quite different in these aspects.
3. Therefore, our subsequent analysis will be based on the results of the following country segmentation.

In [6]:
def process_text(origin_text):
    new_text = origin_text.capitalize()
    new_text = ''.join(new_text.split('-'))
    new_text = ''.join(new_text.split(' '))
    return new_text

In [7]:
def process_country_name(Filename, newFile_path, column):
    df = pd.read_csv(Filename, header=None, sep=',')
    print(df.shape)
    for row in df.iterrows():
        country_name = row[1][column]
        row[1][column]=process_text(country_name)
    df.to_csv(newFile_path, header=None, index=False, sep=',')

In [8]:
def dict_country_code(inputcountry):
    with open(Countries_list_PATH,'r',encoding="utf-8") as f:
        reader = csv.reader(f)
        fieldnames = next(reader) #get first row，as key for the dictionary get from generator next
#         print("fieldnames", fieldnames)
        csv_reader = csv.DictReader(f,fieldnames=fieldnames) #self._fieldnames = fieldnames   # list of keys for the dict 以list的形式存放键名
        for row in csv_reader:
#             print(row)
            d = {}
            for k,v in row.items():
                d[k]=v
            if d['Country']==inputcountry:
                outputcode = d['Code']
#                 print("Succeed! Replaced with", outputcode)
                return(outputcode)
#         print("Error! Cannot find country", inputcountry)
        return False

In [9]:
def dict_country_region(inputcountry):
    with open(Countries_list_PATH,'r',encoding="utf-8") as f:
        reader = csv.reader(f)
        fieldnames = next(reader) #get first row，as key for the dictionary get from generator next
#         print("fieldnames", fieldnames)
        csv_reader = csv.DictReader(f,fieldnames=fieldnames) #self._fieldnames = fieldnames   # list of keys for the dict 以list的形式存放键名
        for row in csv_reader:
#             print(row)
            d = {}
            for k,v in row.items():
                d[k]=v
            if d['Country']==inputcountry:
                outputregion = d['Region']
#                 print("Succeed! Replaced with", outputcode)
                return(outputregion)
#         print("Error! Cannot find country", inputcountry)
        return False

In [14]:
def dict_country_religion(inputcountry):
    with open(Countries_list_PATH,'r',encoding="utf-8") as f:
        reader = csv.reader(f)
        fieldnames = next(reader) #get first row，as key for the dictionary get from generator next
#         print("fieldnames", fieldnames)
        csv_reader = csv.DictReader(f,fieldnames=fieldnames) #self._fieldnames = fieldnames   # list of keys for the dict 以list的形式存放键名
        for row in csv_reader:
#             print(row)
            d = {}
            for k,v in row.items():
                d[k]=v
            if d['Country']==inputcountry:
                outputreligion = d['Religion']
#                 print("Succeed! Replaced with", outputcode)
                return(outputreligion)
#         print("Error! Cannot find country", inputcountry)
        return False

In [15]:
# test function dict_country 
print(dict_country_region('Switzerland'))

Western European and Others Group


In [16]:
# # preprocess every file into country code
# def add_country_code(Filename, newFile_path):
#     df = pd.read_csv(Filename, header=None, sep=',')
#     df['CountryCode']='none'
#     # print(df.columns)
#     print(df.shape)
#     success_count = 0
#     for row in df.iterrows():
#         country_name = row[1][0]
#     #     print("Country name:", country_name)
#         outputcode = dict_country_code(country_name)
#         if(outputcode!=False):
#             row[1]['CountryCode']=outputcode
#             success_count = success_count + 1
#     print("Successfully Converted:", success_count)
#     # print(df.head())
#     df.to_csv(newFile_path,index=False,sep=',')

In [17]:
# country_to_code(Age_limits_CSV_FILE_PATH, "./dataprocessed/Age_limits_processed.csv")
# country_to_code(Public_use_CSV_FILE_PATH, "./dataprocessed/Public_use_processed.csv")

In [18]:
# preprocess add region and religion to each row
def add_code_region_religion(Filename, newFile_path):
    df = pd.read_csv(Filename, header=None, sep=',')
    df['CountryCode']='none'
    df['Region']='none'
    df['Religion']='none'
    print(df.shape)
    success_count = 0
    for row in df.iterrows():
        country_name = row[1][0]
        country_name = process_text(country_name)
        outputcode = dict_country_code(country_name)
        outputregion = dict_country_region(country_name)
        outputreligion = dict_country_religion(country_name)
        if outputregion!=False and outputreligion!=False:
            row[1][0]=country_name
            row[1]['CountryCode']=outputcode
            row[1]['Region']=outputregion
            row[1]['Religion']=outputreligion
            success_count = success_count + 1
    print("Successfully Converted:", success_count)
    df.to_csv(newFile_path,index=False,sep=',')

In [19]:
process_country_name(Countries_list_PATH, "./dataprocessed/Country_list_final.csv", 2)

(177, 5)


In [20]:
process_country_name(Age_limits_CSV_FILE_PATH, "./dataprocessed/Age_limits_processed.csv", 0)

(196, 8)


In [21]:
add_code_region_religion("./dataprocessed/Age_limits_processed.csv", "./dataprocessed/Age_limits_processed.csv")

(196, 11)
Successfully Converted: 162


In [22]:
def text_to_bool():
    df = pd.read_csv(Filename, header=None, sep=',')

In [23]:
add_code_region_religion(Age_limits_CSV_FILE_PATH, "./dataprocessed/Age_limits_processed.csv")

(196, 11)
Successfully Converted: 162


# 2. DataSets for Specific year 2016

In [175]:
from sklearn.preprocessing import MinMaxScaler

'''
use this only when column name is split between the 2 forst rows
merges the two first rows of the csv table into new column name
'''
def merge_col_names(input_path, output_path, row = [0,1]): # column name are in row 1 and 2
    df = pd.read_csv(input_path, header=None, sep=',')
    new_col_names=[]
    for column_name, _ in df.iteritems():
        # checks if the names are different
        if not df.loc[row[0]][column_name] == df.loc[row[1]][column_name]:
            new_col_names.append(df.loc[row[0]][column_name] + ', ' + df.loc[row[1]][column_name])
        else:
            new_col_names.append(df.loc[row[0]][column_name])
    df.columns = new_col_names
    print(new_col_names)
    df=df.drop(labels=0, axis='rows')
    df=df.drop(labels=1, axis='rows')
    df.to_csv(output_path, sep=',', header=True, index=False, mode='w')

In [141]:
## PREPROCESSING AGE LIMITS

## output
# dataframe 1: age_limit, contains the whole data
# dataframe 2: age_limit_num, after dropping missing data

## list of heuristics
HEURISTIC_TOTAL_BAN = 50

# merges column name
merge_col_names(Age_limits_CSV_FILE_PATH, 'dataprocessed/age_limits_processed')

# reads
age_limit = pd.read_csv('dataprocessed/age_limits_processed')

# creates short-cut for column with data specific to age limit
columns_data = list(age_limit.columns)[2:] # without county and year

# lists unique values in the data frame
val1 = pd.Series()
for column in age_limit.columns[2:]:
    val2 = pd.Series(age_limit[column].unique())
    val1 = val1.append(val2)
value_span = list(val1.unique())
print('List of unique values in this data frame: ', value_span)

# sanitizes age_limit
age_limit = age_limit.replace({'total ban': 'Total ban', 
                               'subnational': 'Subnational'})
print('Shape before removing non-numeric values', age_limit.shape)

# puts all values to numbers or NaN
age_limit_num = age_limit.replace({'Total ban': HEURISTIC_TOTAL_BAN,
                                   'None':0,
                                   'Subnational': np.nan,
                                   'No data': np.nan})

# drops rows if it contains NaN
age_limit_num = age_limit_num.dropna(how='any')
print('Shape after removing non-numeric values', age_limit_num.shape)
age_limit_num = age_limit_num.reset_index(drop=True)

# sets the dataframe specific data to float
age_limit_num[columns_data] = age_limit_num[columns_data].astype(dtype=float)

# scales the dataframe specific data  between 0 and 1
mm_scaler = MinMaxScaler(feature_range=(0,1))
df_mm = mm_scaler.fit_transform(age_limit_num[columns_data])
age_limit_num[columns_data] = pd.DataFrame(df_mm)

# to csv
age_limit_num.to_csv('dataprocessed/Age_limits_processed.csv', 
                     sep=',', 
                     header=True,
                     index=False,
                     mode='w')

age_limit_num.head()

['Country', 'Year', 'Age limits off-premise sales,  Beer', 'Age limits off-premise sales,  Wine', 'Age limits off-premise sales,  Spirits', 'Age limits on-premise service,  Beer', 'Age limits on-premise service,  Wine', 'Age limits on-premise service,  Spirits']
List of unique values in this data frame:  ['Total ban', '18', 'None', 'Subnational', 'No data', '16', 'subnational', '17', '21', '25', '20', '15', 'total ban', '19', '13', '14']
Shape before removing non-numeric values (194, 8)
Shape after removing non-numeric values (170, 8)


Country  Year  Age limits off-premise sales,  Beer  \
0  Afghanistan  2016                                 1.00   
1      Albania  2016                                 0.36   
2      Algeria  2016                                 0.36   
3      Andorra  2016                                 0.36   
4       Angola  2016                                 0.36   

   Age limits off-premise sales,  Wine  \
0                                 1.00   
1                                 0.36   
2                                 0.36   
3                                 0.36   
4                                 0.36   

   Age limits off-premise sales,  Spirits  \
0                                    1.00   
1                                    0.36   
2                                    0.36   
3                                    0.36   
4                                    0.36   

   Age limits on-premise service,  Beer  Age limits on-premise service,  Wine  \
0                                  1.00                                  1.00   
1                                  0.36                                  0.36   
2                                  0.36                                  0.36   
3                                  0.36                                  0.36   
4                                  0.36                                  0.36   

   Age limits on-premise service,  Spirits  
0                                     1.00  
1                                     0.36  
2                                     0.36  
3                                     0.36  
4                                     0.36

## 2.1 DataSets Overview

**DataSet 1. Pure alcohol consumption, litres per capita, age 15+**

* Indicator full name: Pure alcohol consumption, litres per capita, age 15+

* Unit: litres per capita

* Dimensions:

* Country (COUNTRY) 

* Supranational group of countries (COUNTRY_GRP)

* Sex (SEX)

* Year of measure (YEAR)

* Years data is available: 1970—2015

In [ ]:
from IPython.core.display import display, HTML
display(HTML('<iframe frameborder="0" class="whoWidget" onload="(function (){if(!document.getElementById(\'ifr\')){var s=document.createElement(\'script\');s.type=\'text/javascript\';s.id=\'ifr\';s.src=\'https://gateway.euro.who.int/Scripts/libs/iframeResizer/iframeResizer.min.js\';document.getElementsByTagName(\'head\')[0].appendChild(s);}})();" src="https://gateway.euro.who.int/en/indicators/visualization-embed/?indId=18314&visId=19443" style="width: 100%;"></iframe>'))

**DataSet 2.Age limits - Alcohol service/sales by country**

* Data type: Count
* Topic: Risk factors
* Rationale: One first needs to be aware of a country's laws and policies regarding alcohol before one can establish whether these influence the level and patterns of alcohol consumption in the population. Thus, the legal status of alcohol production, distribution and consumption is an important indicator.
* Definition:The legal age limit for on-premise service of alcoholic beverages is defined as the age at which a person can be served alcoholic beverages on premises in a country, i.e. alcoholic beverages cannot be served to a person under this age. The legal age limit for off-premise sales of alcoholic beverages is defined as the age at which a person can be sold alcoholic beverages off premises in a country, i.e. alcoholic beverages cannot be sold to a person under this age.
* Disaggregation: Alcoholic beverage type
* Method of estimation: The national authorities of a given country respond to the Global Survey on Alcohol and Health.
* Expected frequency of data dissemination:Periodic
* Expected frequency of data collection:Periodic

Short name: Age limits on-/off-premise sales

Short name:Age limits on-/off-premise service

Total countries:194

In [ ]:
from IPython.core.display import display, HTML
display(HTML('<iframe id=\'rightframe\' name=\'leftiframe\' width="70%" src=\'http://apps.who.int/gho/data/node.main.A1144\'></iframe><br>'))

**DataSet 3.License requirement for alcohol production and retail sales**

* Short name: Licensing
* Data type: Categorical
* Topic: Risk factors
* Rationale: One first needs to be aware of a country's laws and policies regarding alcohol before one can establish whether these influence the level and patterns of alcohol consumption in the population. Thus, the legal status of alcohol production, distribution and consumption is an important indicator.
* Definition: Licensing is defined as the partial government control of production, and sale, wholesale and distribution of alcoholic beverages which requires licensing, as well as of import and export.
* Disaggregation: Alcoholic beverage type
* Method of estimation: The national authorities of a given country respond to the Global Survey on Alcohol and Health.
* Expected frequency of data dissemination:Periodic
* Expected frequency of data collection:Periodic

Retail sales by country

In [ ]:
display(HTML('<iframe id=\'rightframe\' name=\'leftiframe\' width="70%" src=\'http://apps.who.int/gho/data/node.main.A1179\'></iframe><br>'))

Production by country

In [ ]:
display(HTML('<iframe id=\'rightframe\' name=\'leftiframe\' width="70%" src=\'http://apps.who.int/gho/data/node.main.A1178\'></iframe><br>'))

## 2.1 Age-limit v.s. Pure alcohol consumption

In [ ]:
df = pd.read_csv("./dataprocessed/Consumption_processed.csv", names=['Country', '2015-2017', '2009-2011', 'Age limits Beer'])
print(df.head(10))
print(df.columns)

In [ ]:
df.dtype

In [ ]:
df_test = df[2:12]
plt.bar(df_test['Country'], df_test['2015-2017']) # plot
plt.xticks(rotation=30)

In [ ]:
# name_list = ['Monday','Tuesday','Friday','Sunday']
# num_list = [1.5,0.6,7.8,6]
# num_list1 = [1,2,3,1]
# x =list(range(len(num_list)))
total_width, n = 0.8, 2
width = total_width / n
 
plt.bar(df_test['Country'], df_test['2015-2017'], width=width, label='consumption',fc = 'y')
for item in df_test['Country']:
    item = item + width
plt.bar(x, df_test['Age limits Beer'], width=width, label='Age_limits',tick_label = df_test['Country'],fc = 'r')
plt.xticks(rotation=30)
plt.legend()
plt.show()